psycopg_pool/psycopg_pool/pool.py

```diff
@@ -275,30 +278,41 @@ def close(self, timeout: float = 5.0) -> None:
             # Take waiting client and pool connections out of the state
             waiting = list(self._waiting)
             self._waiting.clear()
-            pool = list(self._pool)
+            connections = list(self._pool)
             self._pool.clear()
 
         # Now that the flag _closed is set, getconn will fail immediately,
         # putconn will just close the returned connection.
+        self._stop_workers(waiting, connections, timeout)
+
+    def _stop_workers(
+        self,
+        waiting_clients: Sequence["WaitingClient"] = (),
+        connections: Sequence[Connection[Any]] = (),
+        timeout: float = 0,
+    ) -> None:
 
         # Stop the scheduler
         self._sched.enter(0, None)
 
         # Stop the worker threads
+        workers = self._workers[:]
```

Comment:
This transferring of the workers from the object state to local variables is all over the places here:

- if you took a copy of workers here, I would clear the state immediately, and iterate on the local copy
- the sched runner too, is used from the state then set to none many more lines later

I would clean up this code along the guideline of "if it crashes in any place, the state shouldn't be left inconsistent". Like:
```
workers, self_workers = self._workers, []
sched_runner, self._sched_runner = self._sched_runner, None
```
early in the function, then only use local variable. The above is not really atomic but the intention is clearer :)

Reply:
Done.